gluster-exporter/metric_volume.go

```diff
@@ -311,17 +319,45 @@ func healCounts() error {
 
 	for _, volume := range volumes {
 		name := volume.Name
-		heals, err := gluster.HealInfo(name)
-		if err != nil {
-			log.WithError(err).WithFields(log.Fields{
-				"volume": name,
-			}).Error("Error getting heal info")
-
-			continue
+		if strings.Contains(volume.Type, "Replicate") {
```

Comment:
We could write something like below to reduce the code duplication,
```go
	for _, volume := range volumes {
		name := volume.Name
		// myCommFunc is a function literal, which takes
		// Arg 1. f1    -> a function which takes a string and returns ([]HealEntry, error)
		// (can be 'HealInfo' or 'SplitBrainHealInfo')
		// Arg 2. gVect -> a GaugeVec
		// (can be either 'glusterVolumeHealCount' or 'glusterVolumeSplitBrainHealCount')
		// Arg 3. volName  -> a string representing the volume name
		// Arg 4. errStr   -> the error string in case of error
		myCommFunc := func(f1 func(string) ([]glusterutils.HealEntry, error),
			gVect *prometheus.GaugeVec,
			volName string,
			errStr string) {
			// Get heal count
			heals, err := f1(volName)
			if err != nil {
				log.WithError(err).WithFields(log.Fields{
					"volume": volName,
				}).Error(errStr)
				return
			}
			for _, healinfo := range heals {
				labels := getVolumeHealLabels(volName, healinfo.Hostname, healinfo.Brick)
				gVect.With(labels).Set(float64(healinfo.NumHealEntries))
			}
		}
		if strings.Contains(volume.Type, "Replicate") {
			myCommFunc(gluster.HealInfo,
				glusterVolumeHealCount,
				name, "Error getting heal info")
			myCommFunc(gluster.SplitBrainHealInfo,
				glusterVolumeSplitBrainHealCount,
				name, "Error getting split brain heal info")
		} else if strings.Contains(volume.Type, "Disperse") {
			myCommFunc(gluster.HealInfo,
				glusterVolumeHealCount,
				name, "Error getting heal info")
		}
	}
```
You may have to test it. This is just a suggestion, if the above looks too complex please ignore...

Reply:
Ack. Will try this.

Reply:
We could take the ```myCommFunc``` definition before the ```for``` loop. We don't have to re-declare the same, every time the loop runs.